features/steps/validation_handling.py

```diff
@@ -0,0 +1,281 @@
+            return layer[i]
+    return None
+
+def get_activation_instances(context):
+    """Returns the activation instances of the current context. To be used for 'attribute stacking', e.g. in GEM004"""
+    stack_tree = get_stack_tree(context)
+    return [check_layer_for_entity_instance(i, stack_tree) for i in range(len(stack_tree[0]))]
+
+def validate_step(step_text):
+    def wrapped_step(func):
+        return step(step_text)(execute_step(func))
+
+    return wrapped_step
+
+def execute_step(fn):
+    @wraps(fn)
+    #@todo gh break function down into smaller functions
+    def inner(context, **kwargs):
+        step_type = context.step.step_type
+        if step_type.lower() == 'given': # behave prefers lowercase, but accepts both
+            try:
+                next(fn(context, **kwargs), None)
+            except TypeError:
+                pass
+        elif step_type.lower() == 'then':
+            if not getattr(context, 'applicable', True):
+                validation_outcome = ValidationOutcome(
+                    outcome_code=ValidationOutcomeCode.N00010,  # "NOT_APPLICABLE", Given statement with schema/mvd check failed
+                    observed=None,
+                    expected=None,
+                    feature=context.feature.name,
+                    feature_version=misc.define_feature_version(context),
+                    severity=OutcomeSeverity.NA,
+                    check_execution_id=check_execution_id
+                )
+                context.gherkin_outcomes.append(validation_outcome)
+            else:
+                instances = getattr(context, 'instances', None) or (context.model.by_type(kwargs.get('entity')) if 'entity' in kwargs else [])
+
+                #exclude empty list from instances or no stack_tree; e.g. in case of SPS001
+                try:
+                    activation_instances = get_activation_instances(context) if instances and get_stack_tree(context) else instances
+                except TypeError: #e.g. with IFC001
+                    activation_instances = instances
+
+
+                validation_outcome = ValidationOutcome(
+                    outcome_code=ValidationOutcomeCode.X00040,  # "Executed", but not no error/pass/warning
+                    observed=None,
+                    expected=None,
+                    feature=context.feature.name,
+                    feature_version=misc.define_feature_version(context),
+                    severity=OutcomeSeverity.EXECUTED,
+                    check_execution_id=check_execution_id
+                )
+                context.gherkin_outcomes.append(validation_outcome)
+
+                for i, inst in enumerate(instances):
+                    activation_inst = inst if activation_instances==instances else activation_instances[i]
+                    step_results = list(fn(context, inst = inst, **kwargs)) # note that 'inst' has to be a keyword argument
+                    for result in step_results:
```

Comment:
The reason, for me, is more of a preference then a technicality. I find that, when using keyword arguments, code is often clearer. For instance, in the step_implementation we're also yielding keyword arguments instead of positional.
`yield StepResult(expected=representation_type, observed=None)`
 When writing other step implementations or making adjustments, I find it clear what the role of a variable `representation_type` is in the following handling. More so than in the case of writing positional arguments `yield StepResult(representation_type, None)`, as we did before. 
 
Beside clarity (inst can always be retrieved by a simple `kwargs.get(inst, None)` returning the inst in 'Then' step_implementations and 'None' in case of 'Given' ) positional arguments don't work when one step_impl is written for both 'Then' and 'Given'. In the first, three arguments are passed to the function and, in the latter, there are two. 
I guess a way around this would be something like writing step_impl(context, variable1, variable2, *args) and then passing `inst` as a positional argument. But then, as mentioned, I found using keyword arguments clearer.